doc/dev/python_scheduling/autotune_inner_reduction.py

```diff
@@ -0,0 +1,395 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+# Owner(s): ["module: nvfuser"]
+
+import torch
+import itertools
+from nvfuser import FusionDefinition, SchedulerType, DataType, ParallelType
+from enum import Enum
+from dataclasses import dataclass
+
+from autotune_utils import (
+    ScriptConfiguration,
+    collect_data,
+    separate_data,
+    test_model_rmse,
+    test_model,
+    ceil_div,
+    floor_div,
+)
+
+
+# ================================ Description ================================
+
+# This script defines four inner reduction fusions:
+#
+# 1. Inner Sum
+#    y = sum(x, dim=-1)
+#
+# 2. Add Sum
+#    z = sum(x1 + x2 + x3 + x4, dim=-1)
+#
+# 3. Tanh Sum
+#    y = sum(tanh(x), dim=-1)
+#
+# 4. Exp Sum
+#    z = sum(exp(x), dim=-1)
+#
+# Script Sequence:
+#
+# 1. Define a nvfuser fusion and its pytorch eager mode reference.
+#
+# 2. Profile the CUDA kernel performance by iterating over a set of input
+# arguments and scheduler configurations.
+#
+# 3. Train a regression model to predict the desired performance metric given
+# some input arguments and a scheduler configuration.
+#
+# 4. Measure the performance of the regression model.
+#  - Calculate RMSE of predicted and actual performance on test set.
+#  - Find the configuration with the best performance using regression model.
+#    Then, compare against the heuristic configuration selected by nvfuser.
+#  - For a specific batch size, gather performance across a range of hidden
+#    sizes. Calculate performance for best predicted and nvfuser
+#    configurations. Plot a chart comparing performance using matplotlib.
+#
+# The selected performance metric is effective_bandwidth_gbs. The empirical
+# scheduler selects the configuration that has the highest predicted
+# effective_bandwidth_gbs.
+
+# =============================================================================
+
+
+class AutotuneInnerReduction:
+    class FUSION(Enum):
+        INNER_SUM = 1
+        ADD_SUM = 2
+        TANH_SUM = 3
+        EXP_SUM = 4
+
+    @dataclass(unsafe_hash=True)
+    class InnerReductionConfiguration:
```

Comment:
add comment about how the generated kernel changes when different paras are used.